polaris/polaris/management/commands/process_pending_deposits.py

```diff
@@ -729,14 +866,185 @@ async def save_as_pending_signatures(cls, transaction, server):
+                        datetime.timezone.utc
+                    )
+                    heartbeat.save()
+                    logger.info(
+                        f"lock on key: {PROCESS_PENDING_DEPOSITS_LOCK_KEY} created"
+                    )
+                    return True
+                delta = (
+                    datetime.datetime.now(datetime.timezone.utc)
+                    - heartbeat.last_heartbeat
+                )
+                logger.info(f"lock delta: {delta.seconds} seconds")
+                if delta.seconds > heartbeat_interval * 5:
+                    heartbeat.last_heartbeat = datetime.datetime.now(
+                        datetime.timezone.utc
+                    )
+                    heartbeat.save()
+                    logger.info(
+                        f"lock on key: {PROCESS_PENDING_DEPOSITS_LOCK_KEY} acquired"
+                    )
+                    return True
+            logger.info(
+                f"unable to acquire lock on key: {key}, retrying in {heartbeat_interval} seconds..."
+            )
+            attempt += 1
+            time.sleep(heartbeat_interval)
+
+
+class Command(BaseCommand):
+    """    
+    This process handles all of the transaction submission logic for deposit transactions.
+    When this command is invoked, Polaris queries the database for transactions in the
+    following scenarios and processes them accordingly.
+    A transaction is in the ``pending_user_transfer_start`` or ``pending_external`` status.
+        Polaris passes these transaction the
+        :meth:`~polaris.integrations.RailsIntegration.poll_pending_deposits` integration
+        function, and the anchor is expected to return :class:`~polaris.models.Transaction`
+        objects whose funds have been received off-chain. Polaris then checks if each
+        transaction is in one of the secenarios outlined below, and if not, submits the
+        return transactions them to the Stellar network. See the
+        :meth:`~polaris.integrations.RailsIntegration.poll_pending_deposits()` integration
+        function for more details.
+    A transaction’s destination account does not have a trustline to the requested asset.
+        Polaris checks if the trustline has been established. If it has, and the transaction’s
+        source account doesn’t require multiple signatures, Polaris will submit the transaction
+        to the Stellar Network.
+    **Optional arguments:**
+        -h, --help            show this help message and exit
+        --interval INTERVAL, -i INTERVAL
+                              The number of seconds to wait before restarting
+                              command. Defaults to 10.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        signal.signal(signal.SIGINT, self.exit_gracefully)
+        signal.signal(signal.SIGTERM, self.exit_gracefully)
+
+    @staticmethod
+    async def exit_gracefully():  # pragma: no cover
```

Comment:
Does asyncio implicitly cancel tasks when it receives SIGTERM? I'm trying to understand what happens to the event loop when a signal is sent, because right now it just seems like the process will continue on after deleting the heartbeat key until the OS just kills it with SIGKILL.

Reply:
still testing this but i'm planning on adding:
`        tasks = [task for task in asyncio.Task.all_tasks() if task is not
                asyncio.tasks.Task.current_task()]
        list(map(lambda task: task.cancel(), tasks))
        await asyncio.gather(*tasks, return_exceptions=True)`